### DISCUSSION TITLE
Clickhouse inserts performance degrading over time

### DISCUSSION THREAD

**Discussion by andr0s** [ClickHouse/ClickHouse#68085] [Q&A]
Server: 32G, 3TB SATA soft RAID.

We do tons of inserts, around 1 billion per day. But we want to do more.

We did not touch CH configs much, besides limiting RAM to 20G.

We noticed that inserts speed gets lower and lower.

Brief investigation showed there are constant background merges going on, not sure if it's normal though. Our main table (where we insert all the data into) currently has 13 parts (as reported by `SELECT table, count() AS parts FROM system.parts WHERE active GROUP BY table ORDER BY parts DESC;`).

The data we insert is pretty even, i.e. there are no spikes in size or quantity.

There are no other processes running on this server, that can consume its resources much.

I don't know much about Clickhouse honestly, so forgive me pls if I missed to provide some important information.

I suspect that it all might be related to the merging process going on, and I'm ok to let Clickhouse data take more space than in an ideal world - say, i can sacrifice 30-50% more space, if the inserts performance is high. But again, I'm new to CH so I'm not sure what the real reason of slowdowns is. But we are not even running selects, the data is only getting written, so I expect the database to not do some magic tricks (that are not transparent and hard to debug) and just keep writing the data with decent performance.

```
SELECT elapsed, progress, num_parts, memory_usage, merge_type, merge_algorithm FROM system.merges ORDER BY elapsed DESC;
   ┌────────elapsed─┬────────────progress─┬─num_parts─┬─memory_usage─┬─merge_type─┬─merge_algorithm─┐
1. │ 1624.020750063 │  0.6477225422807932 │         1 │     55218689 │            │                 │
2. │  174.454246703 │ 0.04875030880082744 │         1 │     47125619 │            │                 │
   └────────────────┴─────────────────────┴───────────┴──────────────┴────────────┴─────────────────┘

2 rows in set. Elapsed: 0.002 sec. 
```

BTW we are using latest clickhouse I believe (`ClickHouse client version 24.8.1.1210 (official build).`), all installed via the official .sh script.

Thanks

**Comment by den-crane** [comment]
>We noticed that inserts speed gets lower and lower.
>Brief investigation showed there are constant background merges going on, not sure if it's normal though.

It depends on disk and CPU performance, but it's expected that merges utilize disk and CPU and thus can slow down inserts.
You can reduce merges aggressiveness by several ways.
But for now it's unclear that is the best way of actions because there is no information about the table (DDL), partitioning, number of partitions, parts, average size of parts, number of rows, data retention, how many CPU, disks, etc.

Wild guess: change partitioning to daily, limit the size of part of the table `setting max_bytes_to_merge_at_max_space_in_pool = '30G'`
